Framework/API/src/DetectorSearcher.cpp

```diff
@@ -0,0 +1,224 @@
+#include "MantidAPI/DetectorSearcher.h"
+#include "MantidKernel/ConfigService.h"
+#include "MantidGeometry/Instrument/ReferenceFrame.h"
+
+#include <tuple>
+
+using Mantid::Kernel::V3D;
+using Mantid::Geometry::InstrumentRayTracer;
+using Mantid::Geometry::IDetector;
+using Mantid::Geometry::ReferenceFrame;
+using namespace Mantid;
+using namespace Mantid::API;
+
+/** Create a new DetectorSearcher for the given instrument
+ *
+ * The search strategy will be determined in the constructor based on the
+ * given instrument geometry
+ *
+ * @param instrument :: the instrument to find detectors in
+ * @param detInfo :: the API::DetectorInfo object for this instrument
+ */
+DetectorSearcher::DetectorSearcher(Geometry::Instrument_const_sptr instrument,
+                                   const API::DetectorInfo &detInfo)
+    : m_usingFullRayTrace(instrument->containsRectDetectors() ==
+                          Geometry::Instrument::ContainsState::Full),
+      m_crystallography_convention(Kernel::ConfigService::Instance().getString(
+                                       "Q.convention") == "Crystallography"),
+      m_detInfo(detInfo), m_instrument(instrument) {
+  // choose the search strategy to use
+  if (!m_usingFullRayTrace) {
+    createDetectorCache();
+  } else {
+    m_rayTracer = Kernel::make_unique<InstrumentRayTracer>(instrument);
+  }
+}
+
+/** Create a NearestNeighbours search tree for the current instrument
+ */
+void DetectorSearcher::createDetectorCache() {
+  std::vector<Eigen::Vector3d> points;
+  points.reserve(m_detInfo.size());
+  m_indexMap.reserve(m_detInfo.size());
+
+  for (size_t pointNo = 0; pointNo < m_detInfo.size(); ++pointNo) {
+    if (m_detInfo.isMonitor(pointNo) || m_detInfo.isMasked(pointNo))
+      continue; // detector is a monitor or masked so don't use
+
+    // calculate a Q vector for each detector
+    // This follows a method similar to that used in IntegrateEllipsoids
+    const auto &det = m_detInfo.detector(pointNo);
+    const auto tt1 = det.getTwoTheta(V3D(0, 0, 0), V3D(0, 0, 1)); // two theta
+    const auto ph1 = det.getPhi();                                // phi
+    auto E1 =
+        V3D(-std::sin(tt1) * std::cos(ph1), -std::sin(tt1) * std::sin(ph1),
+            1. - std::cos(tt1)); // end of trajectory
+    E1 = E1 * (1. / E1.norm());  // normalize
+
+    Eigen::Vector3d point(E1[0], E1[1], E1[2]);
+
+    // Ignore nonsensical points
+    if (point.hasNaN())
+      continue;
```

Comment:
This should only be `NaN` when the computed Q is non physical. For example is the Q value was directly along the beam direction.